### PR TITLE
Support multi sample guides in Trace_ELBO

### PR DESCRIPTION
**Pull request by fehiepsi**
This PR enables running SVI with guides which propose multiple samples.

### CODE REVIEW

File: numpyro/infer/elbo.py [hunk 1]
```diff
@@ -116,15 +124,30 @@ class Trace_ELBO(ELBO):
...
         num_particles-many particles in parallel. If False use `jax.lax.map`.
         Defaults to True.
+    :param multi_sample_guide: Whether to make an assumption that the guide proposes
+        multiple samples.
     """
 
+    def __init__(
+        self, num_particles=1, vectorize_particles=True, multi_sample_guide=False
+    ):
+        self.multi_sample_guide = multi_sample_guide
```
Comment (martinjankowiak) [inline numpyro/infer/elbo.py:134]: what happens if `multi_sample_guide==True` and `num_particles > 1`?

File: numpyro/infer/elbo.py [hunk 2]
```diff
@@ -116,15 +124,30 @@ class Trace_ELBO(ELBO):
     :param vectorize_particles: Whether to use `jax.vmap` to compute ELBOs over the
         num_particles-many particles in parallel. If False use `jax.lax.map`.
         Defaults to True.
+    :param multi_sample_guide: Whether to make an assumption that the guide proposes
```
Comment (martinjankowiak) [inline numpyro/infer/elbo.py:127]: why not infer this from `isinstance(guide, ...)`? do we expect guides with this attribute apart from a small number of `AutoGuide`s we construct?

File: numpyro/infer/elbo.py [hunk 3]
```diff
@@ -135,29 +158,65 @@ def single_particle_elbo(rng_key):
...
+                model_log_density = model_log_density.sum(0)
+                # log p(z) - log q(z)
+                elbo_particle = (model_log_density - guide_log_density) / seeds.shape[0]
+            else:
+                seeded_model = seed(model, model_seed)
+                replay_model = replay(seeded_model, guide_trace)
+                model_log_density, model_trace = log_density(
+                    replay_model, args, kwargs, params
+                )
+                check_model_guide_match(model_trace, guide_trace)
```
Comment (martinjankowiak) [inline numpyro/infer/elbo.py:198]: so we lose these two checks if `multi_sample_guide==True`?

Comment (fehiepsi) [reply numpyro/infer/elbo.py:201]: we can check the later one. The first one is tricky.

File: numpyro/infer/elbo.py (hunk 2 above)
Comment (fehiepsi) [reply numpyro/infer/elbo.py:127]: please feel free to remove this when such auto guides are available.

File: numpyro/infer/elbo.py (hunk 1 above)
Comment (fehiepsi) [reply numpyro/infer/elbo.py:134]: num_particles behaves independently. i.e. we might have num_particles x num_samples_from_guide samples.

**Review by martinjankowiak — APPROVED** [review]

**Pull request merged**